smarts-imitation/smarts_imitation/envs/core.py

```diff
@@ -247,23 +295,149 @@ def _init_scenario(self):
+        rows = list(rows)
+        if len(rows) == 0:
+            print(
+                f"Warning: can't get current vehicle history position: id {vehicle_id}"
+            )
+            return None
+        else:
+            return rows[0]
+
+    def _get_vehicle_final_history_position(self, vehicle_id):
+        if self.time_slice:
+            rounder = rounder_for_dt(self.smarts._last_dt)
+            end_time = rounder(self.vehicle_end_times[vehicle_id])
+            start_time = end_time - self.smarts._last_dt
+            rows = self._vehicle_pos_between(vehicle_id, start_time, end_time)
+            rows = list(rows)
+            assert len(rows) > 0, f"final rows: {rows}"
+            return rows[0]
+        else:
+            rows = self.scenario._traffic_history.vehicle_trajectory(vehicle_id)
+            rows = list(rows)
+            assert len(rows) > 0
+            tr = rows[-1]
+            final_position = VehiclePosition(tr.position_x, tr.position_y, None)
+            return final_position
+
+    def _discover_sliced_vehicle_missions(self):
+        """Retrieves the missions of traffic history vehicles."""
+        vehicle_missions = {}
+        for v in self.vehicles:
+            vehicle_id = v.vehicle_id
+            start_time = self.vehicle_start_times[vehicle_id]
+            rows = self._vehicle_pos_between(
+                vehicle_id,
+                start_time=start_time - 0.1,
+                end_time=start_time,
+            )
+            rows = list(rows)
+            assert len(rows) > 0, rows
+            sim_time = rows[0].sim_time
+            start, speed = self.scenario.get_vehicle_start_at_time(vehicle_id, sim_time)
+            entry_tactic = default_entry_tactic(speed)
+            veh_config_type = self.scenario._traffic_history.vehicle_config_type(
+                vehicle_id
+            )
+            veh_dims = self.scenario._traffic_history.vehicle_dims(vehicle_id)
+            vehicle_missions[vehicle_id] = Mission(
+                start=start,
+                entry_tactic=entry_tactic,
+                goal=TraverseGoal(self.scenario.road_map),
+                start_time=sim_time,
+                vehicle_spec=VehicleSpec(
+                    veh_id=vehicle_id,
+                    veh_config_type=veh_config_type,
+                    dimensions=veh_dims,
+                ),
+            )
+
+        return vehicle_missions
+
+    def _init_vehicle_missions(self):
+
+        if self.vehicles is None:
+            self.time_slice = False
+            self.vehicle_missions = (
+                self.scenario.discover_missions_of_traffic_histories()
+            )
+            self.vehicle_ids = list(self.vehicle_missions.keys())
+            self.vehicle_start_times = {
+                v_id: self.vehicle_missions[v_id].start_time
+                for v_id in self.vehicle_ids
+            }
+        elif self.vehicles[0].start_time is None:
```

Comment:
I wonder in what cases self.vehicles[0].start_time is None or not.

Reply:
I have added more comments.

Reply:
So why self.vehicles[0].start_time can be None? It seems to be the time that the vehicle first appears.

Reply:
In the case that self.vehicles[0].start_time is `None`, user does not specify the `start_time` and `end_time` of each vehicle. So the default start time is used (first seen time) to initialize the vehicles. 

In fact, the `self.vehicles[0].start_time` and `self.vehicles[0].end_time` is only used for the `cut-in cases training` now. (I just filter out some cut-in cases from the traffic histories and save the start_time and end_time and the corresponding vehicle_id and replay them during trianing.)

Reply:
Ok, thanks for your explaination.